bindings/azure/openai/openai.go

```diff
@@ -0,0 +1,332 @@
+		if err != nil {
+			return fmt.Errorf("error getting credentials object: %w", err)
+		}
+
+		p.client, err = azopenai.NewClientWithKeyCredential(m.Endpoint, keyCredential, m.DeploymentID, nil)
+		if err != nil {
+			return fmt.Errorf("error creating Azure OpenAI client: %w", err)
+		}
+	} else {
+		// fallback to Azure AD authentication
+		settings, innerErr := azauth.NewEnvironmentSettings(meta.Properties)
+		if innerErr != nil {
+			return fmt.Errorf("error creating environment settings: %w", innerErr)
+		}
+
+		token, innerErr := settings.GetTokenCredential()
+		if innerErr != nil {
+			return fmt.Errorf("error getting token credential: %w", innerErr)
+		}
+
+		p.client, err = azopenai.NewClient(m.Endpoint, token, m.DeploymentID, nil)
+		if err != nil {
+			return fmt.Errorf("error creating Azure OpenAI client: %w", err)
+		}
+	}
+
+	return nil
+}
+
+// Operations returns list of operations supported by OpenAI binding.
+func (p *AzOpenAI) Operations() []bindings.OperationKind {
+	return []bindings.OperationKind{
+		ChatCompletionOperation,
+		CompletionOperation,
+	}
+}
+
+// Invoke handles all invoke operations.
+func (p *AzOpenAI) Invoke(ctx context.Context, req *bindings.InvokeRequest) (resp *bindings.InvokeResponse, err error) {
+	if req == nil || len(req.Metadata) == 0 {
+		return nil, fmt.Errorf("invalid request: metadata is required")
+	}
+
+	startTime := time.Now().UTC()
+	resp = &bindings.InvokeResponse{
+		Metadata: map[string]string{
+			"operation":  string(req.Operation),
+			"start-time": startTime.Format(time.RFC3339Nano),
+		},
+	}
+
+	switch req.Operation { //nolint:exhaustive
+	case CompletionOperation:
+		response, err := p.completion(ctx, req.Data, req.Metadata)
+		if err != nil {
+			return nil, fmt.Errorf("error performing completion: %w", err)
+		}
+		responseAsBytes, _ := json.Marshal(response)
+		resp.Data = responseAsBytes
+
+	case ChatCompletionOperation:
+		response, err := p.chatCompletion(ctx, req.Data, req.Metadata)
+		if err != nil {
+			return nil, fmt.Errorf("error performing chat completion: %w", err)
+		}
+		responseAsBytes, _ := json.Marshal(response)
+		resp.Data = responseAsBytes
+
+	default:
+		return nil, fmt.Errorf(
+			"invalid operation type: %s. Expected %s, %s",
+			req.Operation, CompletionOperation, ChatCompletionOperation,
+		)
+	}
+
+	endTime := time.Now().UTC()
+	resp.Metadata["end-time"] = endTime.Format(time.RFC3339Nano)
+	resp.Metadata["duration"] = endTime.Sub(startTime).String()
+
+	return resp, nil
+}
+
+func (s *ChatSettings) Decode(in any) error {
+	return config.Decode(in, s)
+}
+
+func (p *AzOpenAI) completion(ctx context.Context, message []byte, metadata map[string]string) (response []*azopenai.Choice, err error) {
+	var prompt Prompt
+	err = json.Unmarshal(message, &prompt)
+	if err != nil {
+		return []*azopenai.Choice{}, fmt.Errorf("error unmarshalling the input object: %w", err)
+	}
+
+	messageSettings := make(map[string]any)
+	err = json.Unmarshal(message, &messageSettings)
```

Comment:
Can you add the settings to the Prompt struct, so then you unmarshal only once?

Something like:

```go
type Prompt struct {
	Prompt string `json:"prompt"`
    Temperature      float32 `json:"temperature"`
	MaxTokens        int32   `json:"maxTokens"`
	TopP             float32 `json:"topP"`
	N                int32   `json:"n"`
	PresencePenalty  float32 `json:"presencePenalty"`
	FrequencyPenalty float32 `json:"frequencyPenalty"`
}
```

Then you have a single object prompt that contains both the prompt and the settings... And you don't need ChatSettings anymore either

Reply:
Good idea.